lib/using_restricted_array.rb

```diff
@@ -1,61 +1,122 @@
-require_relative 'restricted_array.rb'
+require_relative "restricted_array.rb"
 # RestrictedArray can be created using a specified size, or a random size in
 # the range of 1-20 will be chosen for you.
 # All values are integers in the range of 1-221.
 # RestrictedArray cannot be resized.
 
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear or O(n), where n is the number of items in the array
+#                  or the length of the input array
+# Space complexity: Constant or O(1) because we are not creating any new data
+#                   structures and therefore not taking up more space in memory
 def length(array)
-  raise NotImplementedError
+  i = 0
+  while array[i] != nil
+    i += 1
+  end
+
+  return i
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear or O(n), where n is the number of items in the array
+#                  or the length of the input array
+# Space complexity: Constant or O(1) because we are not creating any new data
+#                   structures and therefore not taking up more space in memory
 def print_array(array)
-  raise NotImplementedError
+  len = length(array)
+
+  i = 0
+  while i < len
+    print array[i]
+    print " "
```

Comment:
Another way to write this is and combine lines 32 and 33 is `print "#{array[i]} "`